### PR TITLE
Implement WebAuthn Shim for Passkeys tests

### PR DESCRIPTION
**Pull request by nlordell**
This PR adds a `tests/utils/webauthn.ts` module that implements a shim for the Browser WebAuthn API (sometimes conflated with Passkeys).

I verified the implementation by testing both credential creation and assertion (i.e. signing) using a reference WebAuthn/Passkey server implementation (specifically `SimpleWebAuthn` which was suggested on the [Passkeys.dev website](https://passkeys.dev/docs/tools-libraries/libraries/#typescript)).

OMG debugging signature issues is annoying :sob: :sob: :sob:

### CODE REVIEW

File: 4337/.solhintignore [hunk 1]
```diff
@@ -0,0 +1 @@
+contracts/test/FCL/
```
Comment (nlordell) [inline 4337/.solhintignore:1]: unrelated to this PR, just noticed when running `npm run lint` locally.

File: 4337/test/utils/webauthn.ts [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+ *
+ * This implementation is inspired by software authenticators found in the Awesome WebAuthn list [1].
+ *
+ * [1]: <https://github.com/herrjemand/awesome-webauthn#software-authenticators>
+ */
+
+import { p256 } from '@noble/curves/p256'
+import { ethers, BytesLike } from 'ethers'
+import CBOR from 'cbor'
+import { isArrayBuffer } from 'util/types'
```
Comment (mmv08) [inline 4337/test/utils/webauthn.ts:13]: AFAIK this method is only available in Node. Ethers has a similar method which we could use. The motivation is to provide a better reference implementation for in-browser usage.

File: 4337/test/utils/webauthn.ts [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+
+  constructor(
+    public rp: string,
+    public user: Uint8Array,
+  ) {
+    this.pk = p256.utils.normPrivateKeyToScalar(p256.utils.randomPrivateKey())
+    this.id = ethers.dataSlice(ethers.keccak256(ethers.dataSlice(p256.getPublicKey(this.pk, false), 1)), 12)
+  }
+
+  public cosePublicKey(): string {
```
Comment (mmv08) [inline 4337/test/utils/webauthn.ts:94]: Took me a while to find out what cose is; perhaps adding a comment together with the method definition is not a bad idea (I know that the actual implementation links the spec, but I was confused by the method name) https://datatracker.ietf.org/doc/html/rfc8152

File: 4337/test/utils/webauthn.ts [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+ *
+ * @param data data to encode to `base64url`
+ * @returns the `base64url` encoded data as a string.
+ */
+export function base64UrlEncode(data: BytesLike | ArrayBufferLike): string {
+  const bytes = isArrayBuffer(data) ? new Uint8Array(data) : ethers.getBytes(data)
+  return Buffer.from(bytes).toString('base64url')
+}
+
+function b2ab(buf: Buffer): ArrayBuffer {
```
Comment (mmv08) [inline 4337/test/utils/webauthn.ts:242]: Imo it's better to stick to Uint8Array here (buffer extends Uint8Array)

Fun read: https://sindresorhus.com/blog/goodbye-nodejs-buffer

File: 4337/test/webauthn/WebAuthn.spec.ts [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+  }
+
+  const user = {
+    id: ethers.getBytes(ethers.id('chucknorris')),
+    name: 'chucknorris',
+    displayName: 'Chuck Norris',
+  }
+
+  describe('navigator.credentials.create()', () => {
+    it('Safe with 4337 Module Deployment', async () => {
```
Comment (mmv08) [inline 4337/test/webauthn/WebAuthn.spec.ts:34]: Is the test description off, or is the test incomplete?

**Review by mmv08 — APPROVED** [review]

File: 4337/test/utils/webauthn.ts (hunk 3 above)
Comment (nlordell) [reply 4337/test/utils/webauthn.ts:94]: The spec is referenced below twice (for example here: https://github.com/safe-global/safe-modules/blob/26fdff1d5d252962e431fd63ece40fa8fb5d7518/4337/test/utils/webauthn.ts#L101), but I’ll add doc comments referencing the spec again as it wasn’t super clear that those are links to the spec.

File: 4337/test/webauthn/WebAuthn.spec.ts (hunk 5 above)
Comment (nlordell) [reply 4337/test/webauthn/WebAuthn.spec.ts:34]: Its a copy-pasta

File: 4337/test/utils/webauthn.ts (hunk 2 above)
Comment (nlordell) [reply 4337/test/utils/webauthn.ts]: We now don't use `Buffer` anymore :smile: acf4235e71d0535c3b0b8cddeb4bb16ec62aff7b

File: 4337/test/utils/webauthn.ts (hunk 3 above)
Comment (nlordell) [reply 4337/test/utils/webauthn.ts:99]: Added more context to the function documentation.

acf4235e71d0535c3b0b8cddeb4bb16ec62aff7b

File: 4337/test/utils/webauthn.ts (hunk 4 above)
Comment (nlordell) [reply 4337/test/utils/webauthn.ts]: `Buffer` be gone!

acf4235e71d0535c3b0b8cddeb4bb16ec62aff7b

File: 4337/test/webauthn/WebAuthn.spec.ts (hunk 5 above)
Comment (nlordell) [reply 4337/test/webauthn/WebAuthn.spec.ts]: Fixed test descriptions in acf4235e71d0535c3b0b8cddeb4bb16ec62aff7b

**Pull request merged**